AzureMonitor: Fixes metric definition for Azure Storage queue/file/blob/table resources.

aangelisc opened the pull request:
- Nested storage account resources (queues/blobs/tables/files) require metric definition of Microsoft.Storage/storageAccounts
- Update tests accordingly


**What this PR does / why we need it**:

Azure Storage Accounts can contain queue/table/file/blob sub-resources. Currently, we use the metric namespace as the filter for identifying resources. However, the correct metric definition for these resources is `Microsoft.Storage/storageAccounts`. This change appropriately sets the metric definition.

**Which issue(s) this PR fixes**:

Fixes #48794 
<!--

- Automatically closes linked issue when the Pull Request is merged.

Usage: "Fixes #<issue number>", or "Fixes (paste link of issue)"

-->

**Special notes for your reviewer**:

Tested using a storage account with a queue created and ensuring the correct metrics are displayed.

aangelisc committed:
Appropriately set metric definition

- Nested storage account resources (queues/blobs/tables/files) require metric definition of Microsoft.Storage/storageAccounts
- Update tests accordingly

sarahzinger commented on public/app/plugins/datasource/grafana-azure-monitor-datasource/azure_monitor/azure_monitor_datasource.ts:
```diff
@@ -207,9 +207,12 @@ export default class AzureMonitorDatasource extends DataSourceWithBackend<AzureM
     let url =
       `${this.resourcePath}/subscriptions/${subscriptionId}/resourceGroups/${resourceGroup}/resources?` +
-      `$filter=resourceType eq '${metricDefinition}'&` +
+      `$filter=resourceType eq '${validMetricDefinition}'&` +
```
cool stuff, is it possible to write a test for this?

sarahzinger reviewed (approved):
Cool stuff thanks for fixing!

Random mostly unrelated things I have noticed here, that might be worth opening in their own issue or might be easy to fix here I'm not sure:
![Screen Shot 2022-05-18 at 5 09 13 PM](https://user-images.githubusercontent.com/6620164/169156573-729df064-12ff-46f9-a51a-f9158c87eacc.png)
- it's kind of odd here ^ the way we have this duplicated with different casing, do you know what's happening here?
- also I noticed when I tried to make a request for these metrics, I kept getting an error using the 'auto' time grain I'm not sure if that's a new issue or an old one, I'm guessing it's an old issue.

Edit to add: it feels worth adding to changelog to me, it'll show up under bug fixes if you add [Fix to the title](https://github.com/grafana/grafana/blob/7e332d6f97a88562b4016aa3d0cfc2ed177423d2/contribute/merge-pull-request.md) and add the label. This way if there's some kind of funky breaking change here we don't understand, people might be able to find the pr more easily to help us debug.

aangelisc committed:
Restructure getResourceNames test

- Add expect on getResource args

aangelisc commented on public/app/plugins/datasource/grafana-azure-monitor-datasource/azure_monitor/azure_monitor_datasource.test.ts:
```diff
@@ -447,28 +447,28 @@ describe('AzureMonitorDatasource', () => {
               expect(results.length).toEqual(1);
               expect(results[0].text).toEqual('storagetest/default');
               expect(results[0].value).toEqual('storagetest/default');
+              expect(ctx.ds.azureMonitorDatasource.getResource).toHaveBeenCalledWith(
```
@sarahzinger I think this should be sufficient as a test, lmk what you think 😊

aangelisc committed:
Merge branch 'main' into 48794-storage-account-nested-resources

aangelisc committed:
Update to fix issue for new query editor

- Reconstruct resourceUri if the resource is a storage account
- Correctly push storage namespaces as options for metric namespaces
- Filter options appropriately

aangelisc committed:
Fix duplicate options

aangelisc committed:
Fix lint issues

kevinwcyu commented on public/app/plugins/datasource/grafana-azure-monitor-datasource/components/MetricsQueryEditor/setQueryValue.ts:
```diff
@@ -109,6 +109,20 @@ export function setMetricNamespace(query: AzureMonitorQuery, metricNamespace: st
     return query;
   }
 
+  let resourceUri = query.azureMonitor?.resourceUri;
+
+  if (resourceUri && metricNamespace?.includes('Microsoft.Storage/storageAccounts')) {
+    const splitUri = resourceUri.split('/');
+    const accountNameIndex = splitUri.findIndex((item) => item === 'storageAccounts') + 1;
+    const baseUri = splitUri.slice(0, accountNameIndex + 1).join('/');
+    if (metricNamespace === 'Microsoft.Storage/storageAccounts') {
+      resourceUri = baseUri;
+    } else {
+      const subNamespace = metricNamespace.split('/')[2];
+      resourceUri = `${baseUri}/${subNamespace}/default`;
+    }
+  }
+
   return {
```
I was just reading this https://docs.microsoft.com/en-us/azure/storage/common/storage-account-overview#types-of-storage-accounts and https://docs.microsoft.com/en-us/azure/storage/common/storage-account-create?toc=%2Fazure%2Fstorage%2Fblobs%2Ftoc.json&tabs=azure-portal#create-a-storage-account-1 after reading this part of the code and the way I understand it is that the storage account is the resource, which is identified by the resource uri, and the different storage services (queue, blob, etc.) have different metrics which is the metric namespace. So I'm assuming we don't need to modify the resourceUri here.

I think it should still work without this. I also checked the Azure Portal and copied the link to the dashboard where I'm checking the storage account for a queue capacity metric and the resourceUri it uses doesn't include the sub namespace (storage service) part.

I clicked the copy link and used `decodeURIComponent` to see the resource uri (it's under the property resourceMetadata.id) in the decoded string.

<img width="264" alt="Screen Shot 2022-05-19 at 11 30 09 AM" src="https://user-images.githubusercontent.com/19530599/169374520-b40938f4-86f5-492d-89e5-d595d90b0577.png">

aangelisc replied:
I don't think it can, if you take a look at the network requests made when selecting one of the other services besides default you'll see the relative URL will look something like this: 

`"/subscriptions/subId/resourcegroups/rgName/providers/Microsoft.Storage/storageAccounts/acctName/queueServices/default/providers/microsoft.Insights/metrics?timespan=2022-05-19T09:30:00.000Z/2022-05-20T09:30:00.000Z&interval=FULL&metricnames=QueueCapacity&aggregation=average&metricNamespace=microsoft.storage%2Fstorageaccounts%2Fqueueservices&validatedimensions=false&api-version=2019-07-01"`

Similarly, if you take a look at https://docs.microsoft.com/en-us/azure/storage/queues/monitor-queue-storage-reference#resource-logs the `resourceId` field implies that the resource URI needs the service and `default` paths appended. I did test this locally and I couldn't get it working without modifying the resource URI (which is annoying as it also then modifies the display name of the picked resource).

kevinwcyu replied:
Thanks for the explanation. I think it would be worth it to add a comment here to explain why the resourceUri needs to be modified just when it's a storage account.

aangelisc replied:
Have done this, thanks @kevinwcyu!

aangelisc committed:
Add comment explaining URI modification

aangelisc committed:
Merge branch 'main' into 48794-storage-account-nested-resources
